mmcv/ops/csrc/common/cuda/min_area_polygons_cuda.cuh

```diff
@@ -0,0 +1,376 @@
+  while (k_index != max_index) {
+    p_k = p_max;
+    k_index = max_index;
+    for (int i = 1; i < n_poly; i++) {
+      sign = (((double)in_poly[i].x - (double)in_poly[Stack[top1]].x) *
+                  ((double)p_k.y - (double)in_poly[Stack[top1]].y) -
+              ((double)p_k.x - (double)in_poly[Stack[top1]].x) *
+                  ((double)in_poly[i].y - (double)in_poly[Stack[top1]].y));
+      if ((sign > 0) || ((sign == 0) && (dis(in_poly[Stack[top1]], in_poly[i]) >
+                                         dis(in_poly[Stack[top1]], p_k)))) {
+        p_k = in_poly[i];
+        k_index = i;
+      }
+    }
+    top1++;
+    Stack[top1] = k_index;
+  }
+
+  for (int i = 0; i <= top1; i++) {
+    right_point[i] = in_poly[Stack[i]];
+  }
+
+  k_index = 0, Stack[0] = 0, top2 = 0;
+
+  while (k_index != max_index) {
+    p_k = p_max;
+    k_index = max_index;
+    for (int i = 1; i < n_poly; i++) {
+      sign = (((double)in_poly[i].x - (double)in_poly[Stack[top2]].x) *
+                  ((double)p_k.y - (double)in_poly[Stack[top2]].y) -
+              ((double)p_k.x - (double)in_poly[Stack[top2]].x) *
+                  ((double)in_poly[i].y - (double)in_poly[Stack[top2]].y));
+
+      if ((sign < 0) || (sign == 0) && (dis(in_poly[Stack[top2]], in_poly[i]) >
+                                        dis(in_poly[Stack[top2]], p_k))) {
+        p_k = in_poly[i];
+        k_index = i;
+      }
+    }
+    top2++;
+    Stack[top2] = k_index;
+  }
+
+  for (int i = top2 - 1; i >= 0; i--) {
+    left_point[i] = in_poly[Stack[i]];
+  }
+
+  for (int i = 0; i < top1 + top2; i++) {
+    if (i <= top1) {
+      in_poly[i] = right_point[i];
+    } else {
+      in_poly[i] = left_point[top2 - (i - top1)];
+    }
+  }
+  n_poly = top1 + top2;
+  for (int i = 0; i < n_poly; i++) {
+    for (int j = 0; j < n_input; j++) {
+      if (point_same(in_poly[i], input_poly[j])) {
+        points_to_convex_ind[i] = j;
+        break;
+      }
+    }
+  }
+}
+
+__device__ inline void Findminbox(float const *const p, float *minpoints) {
+  Point ps1[MAXN];
+  Point convex[MAXN];
+  float pi = 3.1415926;
+  for (int i = 0; i < 9; i++) {
+    convex[i].x = p[i * 2];
+    convex[i].y = p[i * 2 + 1];
+  }
+  int n_convex = 9;
+  int points_to_convex_ind[9] = {-1, -1, -1, -1, -1, -1, -1, -1, -1};
+  Jarvis_and_index(convex, n_convex, points_to_convex_ind);
+  int n1 = n_convex;
+  for (int i = 0; i < n1; i++) {
+    ps1[i].x = convex[i].x;
+    ps1[i].y = convex[i].y;
+  }
+  ps1[n1].x = convex[0].x;
+  ps1[n1].y = convex[0].y;
+
+  float minbbox[5] = {0};
+  minBoundingRect(ps1, n1 + 1, minbbox);
+  float angle = minbbox[0];
+  float xmin = minbbox[1];
+  float ymin = minbbox[2];
+  float xmax = minbbox[3];
+  float ymax = minbbox[4];
+  float R[2][2];
+
+  R[0][0] = cos(angle);
+  R[0][1] = cos(angle - pi / 2);
+  R[1][0] = cos(angle + pi / 2);
+  R[1][1] = cos(angle);
+
+  float temp[1][2];
+  float points[1][2];
+  points[0][0] = xmax;
```

Comment:
Can you merge these 4 nested loops? Or use 1d array for temp and points.

Reply:
Done.